lib/api/objectPost.js

```diff
@@ -0,0 +1,137 @@
+                        parsedContentLength += chunk.length;
+                        passThroughStream.write(chunk);
+                    });
+
+                    request.fileEventData.file.on('end', () => {
+                        passThroughStream.end();
+                        // Setting the file in the request avoids the need to make changes to createAndStoreObject's
+                        // parameters and thus all it's subsequent calls. This is necessary as the stream used to create
+                        // the object is that of the request directly; something we must work around
+                        // to use the file data produced from the multipart form data.
+                        /* eslint-disable no-param-reassign */
+                        request.fileEventData.file = passThroughStream;
+                        /* eslint-disable no-param-reassign */
+                        // Here parsedContentLength will have the total size of the file
+                        // This is used when calculating the size of the object in createAndStoreObject
+                        request.parsedContentLength = parsedContentLength;
+                        return next();
+                    });
+                    return undefined;
+                },
+                function objectCreateAndStore(next) {
+                    writeContinue(request, request._response);
+                    return createAndStoreObject(request.bucketName,
+                        bucket, request.formData.key, objMD, authInfo, canonicalID, null,
+                        request, false, streamingV4Params, overheadField, log, next);
+                },
+            ], (err, storingResult) => {
+                if (err) {
+                    return callback(err, responseHeaders);
+                }
+                setExpirationHeaders(responseHeaders, {
+                    lifecycleConfig: bucket.getLifecycleConfiguration(),
+                    objectParams: {
+                        key: request.key,
+                        date: storingResult.lastModified,
+                        tags: storingResult.tags,
+                    },
+                });
+                if (storingResult) {
+                    // ETag's hex should always be enclosed in quotes
+                    responseHeaders.ETag = `"${storingResult.contentMD5}"`;
+                }
+                const vcfg = bucket.getVersioningConfiguration();
+                const isVersionedObj = vcfg && vcfg.Status === 'Enabled';
+                if (isVersionedObj) {
+                    if (storingResult && storingResult.versionId) {
+                        responseHeaders['x-amz-version-id'] =
+                            versionIdUtils.encode(storingResult.versionId,
+                                config.versionIdEncodingType);
+                    }
+                }
+                return callback(null, responseHeaders);
```

Comment:
the reponse headers should have a `Location` header

Reply:
The other element present in the doc (Bucket and Key) doesn't seem to be present in the response of aws

Reply:
Added with other params
https://github.com/scality/cloudserver/pull/5601/files#diff-aa7cc4a86866b2bf7ce7b395e7d6ff6672ca9415de1726639e5339d8609a0747R99

Reply:
You can't put them like that on headers, they need to be encoded for url.

Any character can be accepted for the key, it will crash if you put invalid character in headers.

Reply:
2a7d4cf93ab3259d6a9529dc73702718ba7f4d12

Reply:
Yes, aws uses `+` to encode space in url and this `+` should not be urlencoded to `%2B`, but if you have a real `+` in your key, then this one should be encoded to `%2B`.

We might not be compatible with AWS if we don't replicate this behavior but I don't think we already to that anywhere else, so it could be a more general ticket to handle this case everywhere. At least now it's not breaking.

___
Yes `Key` is not present in the reponse, and I did not see the `Bucket` neither, so I guess it's ok for now.
___
Try to reorder your assignments, you moved the comment aboute ETAG

Reply:
Also S3 returns absolute URL with value from Host header

Reply:
I can get the absolute URL from the `host` header in the request. Would that be appropriate?

change with updated tests
92d927339ced597292a0d8187b344a71c3a40d69

Reply:
But bucketName is provided in the `Host`, so if you put it as well here it will be duplicated.

And there is no scheme in the Host so you are missing that as well